macros/src/html.rs

```diff
@@ -249,9 +278,184 @@ impl Element {
+                            span.unstable().start().line,
+                            span.unstable().start().column
+                        );
+                    }
+                    eprintln_msg += &format!(
+                        "<{} {}={}> failed to parse attribute value: {{}}",
+                        name_str, attr_str, lit,
+                    );
+                    #[cfg(not(can_show_location_of_runtime_parse_error))]
+                    {
+                        eprintln_msg += "\nERROR: rebuild with nightly to print source location";
+                    }
+
+                    set_attrs.extend(quote!(
+                        element.attrs.#key = Some(#lit.parse().unwrap_or_else(|err| {
+                            eprintln!(#eprintln_msg, err);
+                            panic!("failed to parse string literal");
+                        }));
+                    ));
+                }
+                value => {
+                    let value = process_value(value);
+                    set_attrs.extend(quote!(
+                        element.attrs.#key = Some(std::convert::Into::into(#value));
+                    ));
+                }
+            }
+        }
+
+        let mut builder = TokenStream::new();
+        builder.extend(quote!(
+            dodrio::builder::ElementBuilder::new(#bump, #tag_name)
+        ));
+
+        for (key, _) in self.attributes.iter() {
+            let key_str = TokenTree::from(Literal::string(&stringify_ident(key)));
+            builder.extend(quote!(
+                .attr(#key_str, dodrio::bumpalo::format!(in &#bump, "{}",
+                      element.attrs.#key.unwrap()).into_bump_str())
+            ));
+        }
+
+        for (key, value) in data_attrs
+            .iter()
+            .map(|(k, v)| (TokenTree::from(Literal::string(&k)), v.clone()))
+        {
+            builder.extend(quote!(
+                .attr(#key, #value.into())
+            ));
+        }
+
+        for (key, value) in events.iter() {
+            let key = TokenTree::from(Literal::string(&stringify_ident(key)));
+            let value = process_value(value);
+            builder.extend(quote!(
+                .on(#key, #value)
+            ));
+        }
+
+        let mut make_req_children = TokenStream::new();
+        let mut arg_list = Vec::new();
+        let mut req_nodes = Vec::new();
+        for (index, child) in req_children.into_iter().enumerate() {
+            let req_child = TokenTree::from(Ident::new(
+                &format!("req_child_{}", index),
+                Span::call_site(),
+            ));
+            let child_node = TokenTree::from(Ident::new(
+                &format!("child_node_{}", index),
+                Span::call_site(),
+            ));
+            make_req_children.extend(quote!(
+                let (#req_child, #child_node) = #child;
+            ));
+            builder.extend(quote!(
+                .child(#child_node)
```

Comment:
If the number of children are statically known, you could generate slightly faster code with `.children([a, b, c, ...])` which uses a statically-sized array instead of dynamically growing a vec in the bump arena.

Alternatively, it may make sense to do

```rust
use dodrio::bumpalo::collections::Vec as BumpVec;

my_builder.children(BumpVec::with_capacity_in(size_hint, bump)
```

before the repeated `.child(...)` calls, so that there is less vec storage doubling.

Same for the event listeners and attributes. This should be a tiny bit faster in general, but the examples largely don't do it because the ergo is a bit funky, but that doesn't matter as much when we are generating code :)

Reply:
It should be trivial to build static arrays with all the children and attributes. I'll have a go.

Reply:
It was. 😊

Reply:
With some caveats... I'm having to build a vector of attributes to be able to omit false boolean attributes, but I know its max size at compile time so there's no risk of storage doubling. I also have to construct the list of children dynamically if it contains code blocks, which can return an unknown amount of children.